VK/simulation/study_1.qmd

```diff
@@ -0,0 +1,232 @@
+                B = B, Psi = Psi, Lambda = Lambda, Theta = Theta))
+}
+```
+
+# Simulation Study 1
+```{r}
+# Define the number of repetitions
+n_reps <- 10  
+
+# Paramter Grid:
+params <- expand.grid(
+  model_type = c("1.1", "1.2", "1.3", "1.4"),
+  N = c(100, 400, 6400),
+  reliability = c(0.3, 0.5, 0.7),
+  method = c("SEM", "SAM"),
+  rep = 1:n_reps
+) %>%
+  mutate(id = row_number())
+ 
+# Set population values:
+B_true <- c(
+    'f3~f1' = 0.1, 'f3~f2' = 0.1, 'f3~f4' = 0.1,
+    'f4~f1' = 0.1, 'f4~f2' = 0.1,
+    'f5~f3' = 0.1, 'f5~f4' = 0.1
+)
+true_values <- list(
+    B = B_true
+)
+
+run_study_1 <- function(params, true_values) {
+  # Set RNG and parallel execution
+  RNGkind("L'Ecuyer-CMRG")
+  plan(multisession, workers = detectCores() - 1)
+
+  # Model syntax
+  model_syntax <- "
+            f1 =~ y1 + y2 + y3
+            f2 =~ y4 + y5 + y6
+            f3 =~ y7 + y8 + y9
+            f4 =~ y10 + y11 + y12
+            f5 =~ y13 + y14 + y15
+
+            f3 ~ f1 + f2 + f4
+            f4 ~ f1 + f2
+            f5 ~ f3 + f4"
+
+  run_analysis <- function(data, model_syntax, method = "SEM") {
+  if (method == "SEM") {
+    tryCatch({
+      fit <- sem(model_syntax, data = as.data.frame(data))
+      if (lavInspect(fit, "converged")) {
+        return(list(fit = fit, error = NULL, converged = TRUE))
+      } else {
+        return(list(fit = NULL, error = "Model did not converge", converged = FALSE))
+      }
+    }, error = function(e) {
+      return(list(fit = NULL, error = e$message, converged = FALSE))
+    })
+  } else if (method == "SAM") {
+    tryCatch({
+      fit <- sam(model_syntax, data = as.data.frame(data), mm.list = list(f1 = "f1", f2 = "f2", f3 = "f3", f4 = "f4", f5 = "f5"))
+      if (lavInspect(fit, "converged")) {
+        return(list(fit = fit, error = NULL, converged = TRUE))
+      } else {
+        return(list(fit = NULL, error = "Model did not converge", converged = FALSE))
+      }
+    }, error = function(e) {
+      return(list(fit = NULL, error = e$message, converged = FALSE))
+    })
+  } else {
+    stop("Unknown method specified")
+  }
+  }
+
+  # Evaluate method performance
+  evaluate_performance <- function(fit, true_values) {
+    if (is.null(fit) || !lavInspect(fit, "converged")) {
+      return(list(Error = "Model did not converge or fit object is null"))
+    }
+    estimates = coef(fit)
+    estimated_paths = estimates[names(estimates) %in% names(true_values$B)]
+    calculate_metrics(estimated_paths, unlist(true_values$B))
+  }
+
+  # Performance Metrics calculation
+  calculate_metrics <- function(estimated, true) {
+    biases = 100 * (estimated - true) / true  # Percentage bias
+    rmse = sqrt(mean((estimated - true)^2, na.rm = TRUE))
+    list(Bias = mean(biases, na.rm = TRUE), RMSE = rmse)
+  }
+
+  # Run the simulations and analysis in parallel
+  results <- future_pmap(params, function(model_type, N, reliability, method, rep, id) {
+    set.seed(12345 + id)
```

Comment:
Since it is not trivial to find out how to do this, I wrote this function up to generate seeds:

```
RNGkind("L'Ecuyer-CMRG")
set.seed(42) 
reps <- 1000
seeds <- purrr::accumulate(seq_len(reps - 1),
                           \(s, x)parallel::nextRNGStream(s),
                           .init = .Random.seed)
```